node/bft/src/helpers/timing.rs

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2019-2025 Provable Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::HashMap, sync::Arc, time::SystemTime};
+
+#[cfg(feature = "locktick")]
+use locktick::parking_lot::RwLock;
+#[cfg(not(feature = "locktick"))]
+use parking_lot::RwLock;
+
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+
+/// Global storage for round-based event data
+static ROUND_EVENTS: Lazy<Arc<RwLock<HashMap<u64, RoundEvents>>>> = Lazy::new(|| Arc::new(RwLock::new(HashMap::new())));
+
+/// Global storage for subdag-based timing data (using lowest/highest rounds from subdag)
+static SUBDAG_TIMINGS: Lazy<Arc<RwLock<HashMap<(u64, u64), SubdagTimings>>>> =
+    Lazy::new(|| Arc::new(RwLock::new(HashMap::new())));
+
+/// Consensus stages that occur per round
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ConsensusStage {
+    ProposalSeen,
+    ProposalCreated,
+    CertificateAdded,
+}
+
+/// Block processing stages that occur per subdag
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum SubdagStage {
+    SubdagProcessing,
+    PrepareAdvanceToNextQuorumBlock,
+    CheckNextBlock,
+    AdvanceToNextBlock,
+}
+
+/// Timing event for a specific consensus stage
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TimingEvent {
+    pub round: u64,
+    pub timestamp: SystemTime,
+    pub event_type: String,
+}
+
+/// Timing data for round-based events
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct RoundEvents {
+    pub round: u64,
+    pub proposal_seen: Vec<TimingEvent>,
+    pub proposal_created: Vec<TimingEvent>,
+    pub certificate_added: Vec<TimingEvent>,
+}
+
+/// Timing data for a specific subdag (identified by lowest and highest rounds)
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SubdagTimings {
+    pub lowest_round: u64,
+    pub highest_round: u64,
+    pub subdag_processing: Option<(SystemTime, Option<SystemTime>)>,
+    pub prepare_advance_to_next_quorum_block: Option<(SystemTime, Option<SystemTime>)>,
+    pub check_next_block: Option<(SystemTime, Option<SystemTime>)>,
+    pub advance_to_next_block: Option<(SystemTime, Option<SystemTime>)>,
+}
+
+/// Combined timing data for export
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TimingSnapshot {
+    pub timestamp: SystemTime,
+    pub round_events: HashMap<String, RoundEvents>,
+    pub subdag_timings: HashMap<String, SubdagTimings>,
```

Comment:
not sure it's worth it to lose the ordering just to have faster queries only in tests